### PR TITLE
Handle both cases with Async and Sync calls

### PR DESCRIPTION
**Pull request by pavelzubarev**
```  
exception: "java.lang.Exception: Batch: GHcGGy - Error when reading from Cognite Data Fusion.
	at com.cognite.beam.io.fn.read.RetrieveItemsBaseFn.processElement(RetrieveItemsBaseFn.java:93)
Caused by: java.util.concurrent.CompletionException: com.google.cloud.storage.StorageException: Read timed out
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:632)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at com.cognite.client.servicesV1.executor.FileBinaryRequestExecutor.lambda$downloadBinaryAsync$0(FileBinaryRequestExecutor.java:245)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.google.cloud.storage.StorageException: Read timed out
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:233)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.getCurrentUploadOffset(HttpStorageRpc.java:809)
```

### CODE REVIEW

File: src/main/java/com/cognite/client/servicesV1/executor/FileBinaryRequestExecutor.java [hunk 1]
```diff
@@ -296,37 +297,41 @@ public FileBinary downloadBinary(Request request) throws Exception {
...
+                            .setContentLength(contentLength)
                             .build();
                 }
             } catch (Exception e) {
                 catchedExceptions.add(e);
-
+                // depends on the execution path either of two may happen:
+                // 1. async call failed and cause is packed with CompletionException
+                // 2. sync call (small file) throws a direct exception
+                final Throwable cause = e instanceof CompletionException ? e.getCause() : e;
```
Comment (kjetilh33) [inline src/main/java/com/cognite/client/servicesV1/executor/FileBinaryRequestExecutor.java:332]: There should not be any async on this code path? This is the sync code that is (potentially) executed on a separate thread, so it is only outside this method that one can observe CompletionExceptions.

File: src/main/java/com/cognite/client/servicesV1/executor/FileBinaryRequestExecutor.java [hunk 2]
```diff
@@ -492,9 +498,12 @@ public ResponseBinary uploadBinary(FileBinary fileBinary, URL targetURL) throws
                         .withApiRetryCounter(apiRetryCounter);
             } catch (Exception e) {
                 catchedExceptions.add(e);
-
+                // depends on the execution path either of two may happen:
+                // 1. async call failed and cause is packed with CompletionException
+                // 2. sync call (small file) throws a direct exception
+                final Throwable cause = e instanceof CompletionException ? e.getCause() : e;
```
Comment (kjetilh33) [inline src/main/java/com/cognite/client/servicesV1/executor/FileBinaryRequestExecutor.java:504]: There should not be any async on this code path? This is the sync code that is (potentially) executed on a separate thread, so it is only outside this method that one can observe CompletionExceptions.

File: src/main/java/com/cognite/client/servicesV1/executor/RequestExecutor.java [hunk 3]
```diff
@@ -243,9 +240,12 @@ public ResponseBinary executeRequest(Request request) throws Exception {
                         .withApiRetryCounter(apiRetryCounter);
             } catch (Exception e) {
                 catchedExceptions.add(e);
-
+                // depends on the execution path either of two may happen:
+                // 1. async call failed and cause is packed with CompletionException
+                // 2. sync call (small file) throws a direct exception
+                final Throwable cause = e instanceof CompletionException ? e.getCause() : e;
                 // if we get a transient error, retry the call
```
Comment (kjetilh33) [inline src/main/java/com/cognite/client/servicesV1/executor/RequestExecutor.java:247]: There should not be any async on this code path? This is the sync code that is (potentially) executed on a separate thread, so it is only outside this method that one can observe CompletionExceptions.

**Review by kjetilh33 — CHANGES_REQUESTED** [review]

File: src/main/java/com/cognite/client/servicesV1/executor/FileBinaryRequestExecutor.java (hunk 1 above)
Comment (pavelzubarev) [reply src/main/java/com/cognite/client/servicesV1/executor/FileBinaryRequestExecutor.java:332]: by "Async" I mean executed somewhere outside that thread, should I use better wording here? I'm not sure I know what is that. On other note: I did get CompletionException here, that's why I was thinking to make a paranoia type of unpacking.

**Comment by pavelzubarev** [comment]
I think my approach here is to be as much as possible on the safe side of things. So even if we should not have it now, but one day later we switch to use threads, that code will capture either of it

File: src/main/java/com/cognite/client/Files.java [hunk 4]
```diff
@@ -1105,24 +1077,34 @@ public Aggregate aggregate(Request requestParameters) throws Exception {
...
 
+        ImmutableList.Builder<Map<String, Object>> insertItemsBuilder = ImmutableList.builder();
+        for (FileMetadata fileMetadata : filesBatch) {
+            // build request object
+            insertItemsBuilder.add(toRequestInsertItem(fileMetadata));
+        }
         // build request object
         Request postSeqBody = addAuthInfo(Request.create()
-                .withRequestParameters(toRequestInsertItem(fileMetadata)));
+                .withItems(insertItemsBuilder.build()));
```
Comment (pavelzubarev) [inline src/main/java/com/cognite/client/Files.java:1097]: I think that brakes writer creation. Files API doesn't support batch create 🤷‍♂️  @kjetilh33

Comment (kjetilh33) [reply src/main/java/com/cognite/client/Files.java:1097]: This was a single item method. Not a batch method.

Comment (kjetilh33) [reply src/main/java/com/cognite/client/Files.java:1097]: Sooo... a batch of fileMetadata got split into single items via "splitAndCreateFileMetadata()" -> single items created by "createFileMetadata".

Comment (pavelzubarev) [reply src/main/java/com/cognite/client/Files.java:1097]: yes I'm working on the fix of it... there is no items.

**Pull request closed without merging**